llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp

```diff
@@ -1014,12 +1014,49 @@ struct AAAMDGPUNoAGPR
 
 const char AAAMDGPUNoAGPR::ID = 0;
 
+static unsigned getMaxNumPreloadArgs(const Function &F, const DataLayout &DL,
+                                     const TargetMachine &TM) {
+  const GCNSubtarget &ST = TM.getSubtarget<GCNSubtarget>(F);
+  unsigned Offset = 0;
+  unsigned ArgsToPreload = 0;
+  for (const auto &Arg : F.args()) {
+    if (Arg.hasByRefAttr())
+      break;
+
+    Type *Ty = Arg.getType();
+    Align ArgAlign = DL.getABITypeAlign(Ty);
+    auto Size = DL.getTypeAllocSize(Ty);
+    Offset = alignTo(Offset, ArgAlign);
+    if (((Offset + Size) / 4) > ST.getMaxNumUserSGPRs())
+      break;
+
+    Offset += Size;
+    ArgsToPreload++;
+  }
+
+  return ArgsToPreload;
+}
+
 static void addPreloadKernArgHint(Function &F, TargetMachine &TM) {
   const GCNSubtarget &ST = TM.getSubtarget<GCNSubtarget>(F);
-  for (unsigned I = 0;
-       I < F.arg_size() &&
-       I < std::min(KernargPreloadCount.getValue(), ST.getMaxNumUserSGPRs());
-       ++I) {
+  if (!ST.hasKernargPreload())
+    return;
+
+  // Enable kernarg preloading by default on GFX940+.
+  size_t PreloadCount;
+  if (KernargPreloadCount.getNumOccurrences() > 0) {
```

Comment:
Should move this to be a pass parameter?

Reply:
What's the benefit of having the logic outside of the pass and using a pass parameter?

I want to keep it so the flag works as before and users can override the defaults. I would still need the cl opt somewhere.

Reply:
cl::opts are evil and should mostly be purged. They're only for debugging and shouldn't be given to users as a configuration mechanism. I don't really see why users would want direct control over something so specific either.

Reply:
If we removed the cl option we would have to give users another way to enable it on gfx90a, but we don't want to enable it by default there.

Reply:
If it really needs to be a user exposed flag, there should be an attribute for it with a corresponding clang flag to set it

Reply:
I really see where you are coming from, and I agree with you in most circumstances, but this CL option has been used for this purpose since downstream prototypes over a year ago. Now, the flag will only be used for a niche case, where 
 some libraries are already using the flag on gfx90a. I'd like to keep it because I don't actually want it to be a user facing option. This shouldn't be exposed to the user and is really only a "user beware" type hidden flag for downstream libraries that are already using it.